### PR TITLE
fix(persist): #451

### LINKED ISSUE #451: onRehydrateStorage argument
in v3.5.1 the persist middleware's `onRehydrateStorage` option provided the entire state as the argument.  

in 3.5.4 only the persisted data is returned in the argument.

I was relying on having the whole state, to use these options (as an example):

```javascript
const persistOptions = {
    name: my-storage',
    whitelist: ['persisted'],
    onRehydrateStorage: () => {
        return (state: FullState) => {
            state.setRehydrated(true);
        };
    },
};
```

Is this a deliberate change?  If so, I'll find another way to keep track of rehydration status.

thanks

### PR DESCRIPTION
**Pull request by AnatoleLucet**
Closes #451

This is to follow-up #437 & #442.

### CODE REVIEW

File: src/middleware.ts [hunk 1]
```diff
@@ -362,21 +371,15 @@ export const persist =
         }
       })
       .then(() => {
-        postRehydrationCallback?.(stateFromStorageInSync, undefined)
+        postRehydrationCallback?.(
+          { ...configResult, ...stateFromStorageInSync },
```
Comment (dai-shi) [inline src/middleware.ts:375]: Yeah, this is something I concerned with "merged" state. Looks good.
One caveat is L375 and L384 are different objects. Would be nice to make it referentially identical.

Comment (AnatoleLucet) [reply src/middleware.ts:375]: good idea

Comment (AnatoleLucet) [reply src/middleware.ts]: there you go :slightly_smiling_face:

File: src/middleware.ts [hunk 2]
```diff
@@ -362,22 +371,12 @@ export const persist =
         }
       })
       .then(() => {
+        stateFromStorageInSync = { ...configResult, ...stateFromStorageInSync }
```
Comment (dai-shi) [inline src/middleware.ts:374]:
```suggestion
        stateFromStorageInSync = stateFromStorageInSync && { ...configResult, ...stateFromStorageInSync }
```
We need this, right?

Comment (AnatoleLucet) [reply src/middleware.ts:374]: image
:wink:

Comment (dai-shi) [reply src/middleware.ts:374]: The point is creating a new object in the case of async: `configResult !== { ...configResult }`. In other words, `|| configResult` does never work.

Comment (AnatoleLucet) [reply src/middleware.ts:374]: Yes it does. When using an async storage

Comment (dai-shi) [reply src/middleware.ts:374]: Oh, I see. In async, it doesn't reach this line before the return. Then, I'm fine with this.

**Review by dai-shi — APPROVED** [review]

**Comment by dai-shi** [comment]
I will merge this later unless @barbogast says something.

**Comment by dai-shi** [comment]
Let's merge this so that we can work on #459 without conflicts.

**Comment by barbogast** [comment]
Sorry, I'm on vacation and missed the notifications.

Post-merge approval from me :-)

**Pull request merged**